rfc/src/rfcs/0009-function-contracts.md

```diff
@@ -0,0 +1,674 @@
+stubbing allows for modular verification, which paves the way for the
+following two ambitious goals.
+
+[^simple-unsoundness]: The main remaining threat to soundness in the use of
+    contracts, as defined in this proposal, is the reliance on user-supplied
+    harnesses for contract checking (explained in item 2 of [user
+    experience](#user-experience)). A more thorough discussion on the dangers
+    and potential remedies can be found in the [future](#future-possibilities)
+    section.
+
+- **Scalability:** Function contracts are sound (over)abstractions of function
+  behavior. By verifiying the contract against its implemetation and
+  subsequently performing caller verification against the (cheaper) abstraction,
+  verification can be modularized, cached and thus scaled.
+- **Unbounded Verification:** The abstraction provided by the contract can be
+  used instead of a recursive call, thus allowing verification of recursive
+  functions.
+
+Function contracts are completely optional with no user impact if unused. The
+newly introduced APIs are entirely backwards compatible.
+
+
+## User Experience
+
+Function contracts provide a verifiable way to specify function behavior. In
+addition, the specified behavior can subsequently be used as an abstraction of
+the function's behavior at call sites.
+
+The lifecycle of a contract is split roughly into three phases: specification,
+verification and stubbing. Which we will explore on this simple example:
+
+```rs
+fn my_div(dividend: u32, divisor: u32) -> u32 {
+  dividend / divisor
+}
+```
+
+1. In the first phase we **specify** the approximation. Kani provides two new
+   annotations: `requires` (preconditions) to describe the expectations this
+   function has as to the calling context and `ensures` (postconditions) which
+   approximates function outputs in terms of function inputs.
+
+   ```rs
+   #[kani::requires(divisor != 0)]
+   #[kani::ensures(result <= dividend)]
+   fn my_div(dividend: u32, divisor: u32) -> u32 {
+     dividend / divisor
+   }
+   ```
+  
+   `requires` here indicates this function expects its `divisor` input to never
+   be 0, or it will not execute correctly (i.e. panic).
+
+   `ensures` puts a bound on the output, relative to the `dividend` input.
+
+   Conditions in contracts are plain Rust expressions which can reference the
+   function arguments and, in case of `ensures`, the result of the function as a
+   special `result` variable (see [open questions](#open-questions) on a
+   discussion about (re)naming). Syntactically Kani supports any Rust
+   expression, including function calls, defining types etc. However they must
+   be side-effect free[^side-effects] or Kani will throw a compile error.
+
+   Multiple `requires` and `ensures` clauses are allowed on the same function,
+   they are implicitly logically conjoined.
+
+
+2. Next, Kani must make sure that the approximation we specified actually holds
```

Comment:
Does kani warn if a contract is used and does not have a corresponding harness ?